### ISSUE TITLE
CPU starvation due to expensive deduplication of hundreds of flows sent during stack link initialization

### ISSUE THREAD

**Issue by grafnu** [faucetsdn/faucet#3281]
In our stacking topology, when we add ~40 hosts to one switch, things slow down enough in faucet that LLDP packets are being missed (5 sec timeout), so that the stack is not initially stable.

In the case of many hosts, during the startup sequence we see the following message:
`Oct 07 14:13:28 faucet.valve INFO     DPID 177 (0xb1) nz-kiwi-t1sw1 LLDP for Port 10 inactive after 17s
`
Which is not there when we only have a few hosts attached. Configuration file is the same in both cases.

Possibly related to previously addressed issue https://github.com/faucetsdn/faucet/issues/2809

**Comment by anarkiwi** [comment]
It would be helpful to know what what is slowing down.

Perhaps you could do some profiling?

**Comment by anarkiwi** [comment]
I suggest some other next steps:

* Troubleshooting what is actually slowing down (is a function in FAUCET? are the LLDP packets really being dropped? You can look at the OF control channel for example).
* Is there anything else on the switches that looks like it could slow them down? Do they have normal/low CPU usage? Are you polling flow tables frequently? etc.

**Comment by grafnu** [comment]
Yup, I'll diagnose and report back when I have a root cause. Just throwing
it out there in case somebody else is seeing the same behavior!

**Comment by grafnu** [comment]
PR https://github.com/faucetsdn/faucet/pull/3284 fixes a lldp timeout bug that enables a configuration workaround for this.  Not a complete fix (doesn't address root cause of slowdown).

**Comment by grafnu** [comment]
This isn't a complete analysis, but what I know so far -- still investigating (profiling). This is testing a setup with ~100 devices connected to a (virtual) switch. Stacking setup is 4x switches with a bunch of interconnects.

* every-time a LLDP link in stack (7) receives a message
* sent from either end (x2)
* _update_stack_link_state gets called, resulting in ~790 of messages.
* prepare_send_flows gets called with ~549 messages
* This takes about ~10s
* most time in valve_of.dedupe_ofmsgs
* In there it converts each message to string and puts in a set.

I'm guessing is that the convert-to-string conversion is very expensive, and is ok at smaller scales, but with larger number of devices it becomes a problem. This has yet to be experimentally validated.

**Comment by grafnu** [comment]
Profiling results that track the hot-spot in question:
```
         3796476 function calls (3676306 primitive calls) in 1.460 seconds

   Ordered by: cumulative time
   List reduced from 29 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.002    0.002    1.460    1.460 /usr/lib/python3.7/site-packages/faucet/valve_of.py:888(<dictcomp>)
      642    0.002    0.000    1.459    0.002 /usr/lib/python3.7/site-packages/ryu/ofproto/ofproto_parser.py:230(__str__)
12328/642    0.019    0.000    1.455    0.002 /usr/lib/python3.7/site-packages/ryu/lib/stringify.py:160(__str__)
12328/642    0.018    0.000    1.452    0.002 {method 'join' of 'str' objects}
63220/8978    0.042    0.000    1.450    0.000 /usr/lib/python3.7/site-packages/ryu/lib/stringify.py:163(<genexpr>)
    61350    0.203    0.000    1.347    0.000 /usr/lib/python3.7/site-packages/ryu/lib/stringify.py:369(obj_python_attrs)
50892/8336    0.019    0.000    1.338    0.000 {built-in method builtins.repr}
    11393    0.362    0.000    0.955    0.000 /usr/lib/python3.7/inspect.py:316(getmembers)
    11393    0.184    0.000    0.184    0.000 {built-in method builtins.dir}
    11393    0.107    0.000    0.168    0.000 {method 'sort' of 'list' objects}
   646925    0.131    0.000    0.131    0.000 {method 'startswith' of 'str' objects}
   669711    0.116    0.000    0.116    0.000 {built-in method builtins.getattr}
   646925    0.062    0.000    0.062    0.000 {method 'add' of 'set' objects}
   646925    0.061    0.000    0.061    0.000 /usr/lib/python3.7/inspect.py:357(<lambda>)
   646925    0.059    0.000    0.059    0.000 {method 'append' of 'list' objects}
    81900    0.038    0.000    0.038    0.000 {built-in method builtins.hasattr}
   178528    0.016    0.000    0.016    0.000 {built-in method builtins.callable}
      641    0.001    0.000    0.009    0.000 /usr/lib/python3.7/site-packages/ryu/ofproto/ofproto_v1_3_parser.py:960(__str__)
    11393    0.004    0.000    0.007    0.000 /usr/lib/python3.7/inspect.py:72(isclass)
    11393    0.004    0.000    0.004    0.000 /usr/lib/python3.7/site-packages/ryu/lib/stringify.py:156(stringify_attrs)

```

**Comment by anarkiwi** [comment]
Why does update link state do any work if the link is already up?

**Comment by grafnu** [comment]
This is observed on system start, so initially all the links are initially down, so it's getting triggered when each link comes up. And they all come up right at the start. (I was wrong in my assertion that it happens x2 for each link, it's only once when it first comes up.)

I don't really understand why 790 messages are generated in this case (from the call to _update_stack_link_state ) -- I'm not sure if that's something to investigate further.

**Comment by anarkiwi** [comment]
OK thanks for clarifying that makes more sense. Hoever, that is a very high number of messages, so definitely that's worth investigating rather than working around.  deduplicate should be more efficient but it should not need to do that much work.

**Comment by anarkiwi** [comment]
Additionally, sending that many flows many cause problems to DPs (ie. making deduplicate more efficient will just move the problem to the switch CPU which is all but guaranteed to be less powerful than the controller). Therefore reducing the number of messages sent will give better results overall.

**Comment by grafnu** [comment]
* One switch (642 flows)
  * 20 flows are non-port specific
  * 44 active (no-stacking) ports
    * 14 flows each
      * 8 port-ACL rules (as defined by us)
      * 1 set-vlan rule
      * 5 faucet-added broadcast-matching rules (4+default)

Which adds up to 600+ flows (I think the 700 came from a setup with more for testing)

I don't now why the 8 ACL rules are being added/changed when the stack comes up, since theoretically they don't change based on the stacking (which is what triggers this).

The 5 faucet-added rules could likely be coalesced, since they are nearly identical except for they just don't explicitly output to their in-port, which is what varies between each port.

**Comment by grafnu** [comment]
Upload of my debugging logs for posterity.

[all ofmsgs for the switch](https://github.com/faucetsdn/faucet/files/3718512/ofmsgs_642.txt)
[messages just for port 8](https://github.com/faucetsdn/faucet/files/3718514/sorted8_ofmsgs.txt)

**Comment by anarkiwi** [comment]
https://github.com/faucetsdn/faucet/pull/3358 is part of the problem - we repeat state init, which expensively re adds all flows over and over from a cold start

**Comment by anarkiwi** [comment]
https://github.com/faucetsdn/faucet/pull/3405 avoids an extra port bounce/flood rule reprogram when the LIVE flag changes, and https://github.com/faucetsdn/faucet/pull/3402 uses the python lru_cache function to make flood rule generation and packet parsing less expensive.

If there are still problems please open a new issue with precise instructions on how to reproduce.
